pkg/nuctl/command/common/patchmanifest.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2023 The Nuclio Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"sync"
+
+	"github.com/nuclio/logger"
+)
+
+type PatchManifest struct {
+	*patchManifest
+	lock sync.Mutex
+}
+
+type patchManifest struct {
+	Success []string                   `json:"success,omitempty"`
+	Skipped []string                   `json:"skipped,omitempty"`
+	Failed  map[string]FailDescription `json:"failed,omitempty"`
```

Comment:
id think of the datastructure as sync.Map
key is function name
values are
"state" (succeeded, skipped, failed)
"reason" (why it failed)
"retryable" (can be redeployed)
"project" (the project the function belongs to)
etc.

That way, you can avoid using mutex everywhere

Reply:
In this case, the computational complexity when searching for fail functions will increase. Imagine a situation where we have 1000 functions, 999 of which completed successfully and only one ended with an error, then we will have to go through the entire list of functions to find that one